src/lib/eap_aka_sim/encode.c

```diff
@@ -701,70 +639,68 @@ static ssize_t encode_rfc_hdr(uint8_t *out, size_t outlen, fr_da_stack_t *da_sta
 	 */
 	da = da_stack->da[depth];
 
-	p += 2;	/* Leave space for attr + len */
+	fr_dbuff_advance(&work_dbuff, 2);
 	if (da->flags.array) {
-		slen = encode_array(p, outlen - (p - out), da_stack, depth, cursor, encoder_ctx);
+		slen = encode_array(&FR_DBUFF_MAX(&work_dbuff, SIM_MAX_ATTRIBUTE_VALUE_LEN - 2),
+				    da_stack, depth, cursor, encoder_ctx);
 	} else {
-		slen = encode_value(p, outlen - (p - out), da_stack, depth, cursor, encoder_ctx);
+		slen = encode_value(&FR_DBUFF_MAX(&work_dbuff, SIM_MAX_ATTRIBUTE_VALUE_LEN - 2),
+				    da_stack, depth, cursor, encoder_ctx);
 	}
 	if (slen <= 0) return slen;
-	p += slen;
 
 	/*
 	 *	Pad value to multiple of 4
 	 */
-	pad_len = ROUND_UP_POW2(p - out, 4) - (p - out);
+	pad_len = ROUND_UP_POW2(fr_dbuff_used(&work_dbuff), 4) - fr_dbuff_used(&work_dbuff);
 	if (pad_len) {
-		CHECK_FREESPACE(end - p, pad_len);
-
-		memset(p, 0, pad_len);
-		p += pad_len;
+		FR_DBUFF_MEMSET_RETURN(&work_dbuff, 0, pad_len);
 	}
 
-	out[0] = da->attr & 0xff;
-	out[1] = (p - out) >> 2;
+	fr_dbuff_bytes_in(&hdr_dbuff, (uint8_t)da->attr,
+			  (uint8_t)(fr_dbuff_used(&work_dbuff) >> 2));
 
-	FR_PROTO_HEX_DUMP(out, p - out, "Done RFC attribute");
+	FR_PROTO_HEX_DUMP(fr_dbuff_start(&work_dbuff), fr_dbuff_used(&work_dbuff), "Done RFC attribute");
 
-	return (p - out);	/* AT + Length + Data */
+	return fr_dbuff_set(dbuff, &work_dbuff);	/* AT + Length + Data */
 }
 
-static inline ssize_t encode_tlv_internal(uint8_t *out, size_t outlen,
+static inline ssize_t encode_tlv_internal(fr_dbuff_t *dbuff,
 					  fr_da_stack_t *da_stack, unsigned int depth,
 					  fr_cursor_t *cursor, void *encoder_ctx)
 {
 	ssize_t			slen;
-	uint8_t			*p = out, *end = p + outlen, *value;
+	fr_dbuff_t		work_dbuff = FR_DBUFF_NO_ADVANCE(dbuff);
+	fr_dbuff_t		value_dbuff;
+	fr_dbuff_marker_t	value_start;
 	VALUE_PAIR const	*vp = fr_cursor_current(cursor);
 	fr_dict_attr_t const	*da = da_stack->da[depth];
 
-	CHECK_FREESPACE(outlen, 2);
-	*p++ = 0;	/* Reserved (0) */
-	*p++ = 0;	/* Reserved (1) */
-	value = p;
+	FR_DBUFF_BYTES_IN_RETURN(&work_dbuff, 0, 0);
+
+	value_dbuff = FR_DBUFF_NO_ADVANCE(&work_dbuff);
+	fr_dbuff_marker(&value_start, &value_dbuff);
 
-	while ((end - p) > 4) {
-		size_t sublen;
+	for(;;) {
 		FR_PROTO_STACK_PRINT(da_stack, depth);
 
 		/*
 		 *	This attribute carries sub-TLVs.  The sub-TLVs
 		 *	can only carry SIM_MAX_ATTRIBUTE_VALUE_LEN bytes of data.
 		 */
-		sublen = end - p;
-		if (sublen > SIM_MAX_ATTRIBUTE_VALUE_LEN) sublen = SIM_MAX_ATTRIBUTE_VALUE_LEN;
 
 		/*
 		 *	Determine the nested type and call the appropriate encoder
 		 */
 		if (da_stack->da[depth + 1]->type == FR_TYPE_TLV) {
-			slen = encode_tlv_hdr(p, sublen, da_stack, depth + 1, cursor, encoder_ctx);
+			slen = encode_tlv_hdr(&FR_DBUFF_MAX(&work_dbuff, SIM_MAX_ATTRIBUTE_VALUE_LEN),
```

Comment:
using FR_SBUFF_MAX like this is wrong.  It should be initialised outside of the loop and used to restrict the maximum length of all children.

Reply:
Hmm.... the pre-dbuff version looked like it was bounding the lengths of the individual children, but it works on unit tests with their sum constrained to that value; changing to work that way.